feat: Added a way to safely access GlobalObjectIdHash via PrefabHashId [MTT-1645]

samlucas-unity opened the pull request:
Added a way to access the GlobalObjectIdHash for use in ConnectionApproval callback.  It checks the Object is actually a prefab before returning the value.

ConnectionApproval Runtime Determined GlobalObjectIdHash Spawn Not Possible [MTT-1645]

## Changelog

- Added a way to access the GlobalObjectIdHash via PrefabHashId for use in the Connection Approval Callback.

samlucas-unity committed:
MTT-5830 Fix
NetworkMetrics were tracking named message received events but it is possible for message handlers to unregister themselves.
We now cache the name of the message handler before invoking it so the Network Metrics can access the name even if its removed.

samlucas-unity committed:
Merge branch 'develop' into MTT-5380

samlucas-unity committed:
Updated CHANGELOF.md to contain details of this fix.

samlucas-unity committed:
Added a way to access the GlobalObjectIdHash via PrefabHashId.

samlucas-unity committed:
Merge branch 'develop' into MTT-5380

samlucas-unity committed:
Updated the PR number

samlucas-unity committed:
Removed commented out code

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -17,6 +17,31 @@ public sealed class NetworkObject : MonoBehaviour
+        /// Gets the Prefab Hash Id of this object if the object is registerd as a prefab otherwise it returns 0
+        /// </summary>
+        [HideInInspector]
+        public uint PrefabHashId
```
```suggestion
        public uint PrefabIdHash
```

it's actually a hash of an ID, not an ID of a hash :)

samlucas-unity replied:
😁 Oh Yeah!

samlucas-unity committed:
Update com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

samlucas-unity committed:
Update CHANGELOG.md

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -17,6 +17,31 @@ public sealed class NetworkObject : MonoBehaviour
+                    }
+                }
+
+                if ( NetworkManager.PrefabHandler.ContainsHandler( GlobalObjectIdHash ) )
```
The standards check issues are due to the 2 whitespaces on line 36 position 21 and line 36 position 67. 
`if (NetworkManager.PrefabHandler.ContainsHandler( GlobalObjectIdHash ))`
👍

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -17,6 +17,31 @@ public sealed class NetworkObject : MonoBehaviour
+        {
+            get
+            {
+                foreach (var prefabList in NetworkManager.NetworkConfig.Prefabs.NetworkPrefabsLists)
```
Can we cache the result of this so it doesn't have to do this for-loop every time it's accessed? This seems like it would be pretty expensive, but as a property, it's not obvious that it's expensive to access.

samlucas-unity replied:
I don't think we can unfortunately. I've added some tests and this code isn't correct. There is a Prefabs list in NetworkManager.NetworkConfig.Prefabs.Prefabs which is the actual list of prefabs we need to chefck against. NetworkPrefabsLists is just the list of saved prefabs, any prefabs added at runtime won't be considered.  Since that list can change we can't rely on a cached value.  Fortunately the use case for this is limited to the ConnectionApproval callback so its unlikely its going to get hit very often.  I will update the PR soon with those changes.

samlucas-unity committed:
Fixed PrefabHashId get. It should have been accessing the Prefabs list
and not the saved NetworkPrefabsLists as they opnly contain saved prefabs.
Also removed checking the PrefabHandler as this isn't correct.

samlucas-unity committed:
Added PrefabHashId tests.

samlucas-unity committed:
Merge branch 'MTT-5380' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into MTT-5380

samlucas-unity committed:
Merge branch 'develop' into MTT-5380

samlucas-unity committed:
Fixed renaming issue.

NoelStephensUnity committed:
Merge branch 'develop' into MTT-5380

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkObject/NetworkObjectPropertyTests.cs:
```diff
@@ -0,0 +1,63 @@
+        protected override void OnServerAndClientsCreated()
+        {
+            // create prefab
+            var gameObject = new GameObject("TestObject");
```
Just FYI:
`CreateNetworkObjectPrefab` handles all of this for you (and then some).

NoelStephensUnity committed:
test

Updating the test for this PR.

NoelStephensUnity committed:
style

removing System.Collections that is not used
